src/js/kits/model-kit-select/ModelKitSelect.js

```diff
@@ -0,0 +1,192 @@
+// TODO: handle allowMultiple yes/no
+import React, { useState, useReducer } from 'react';
+import PropTypes from 'prop-types';
+import { useAsync, useDebounce } from 'react-use';
+
+import { ModelKitConsumer } from '../../data/kits/modelkit';
+import { FilterForm } from './FilterForm';
+import { KitPreview } from './KitPreview';
+
+const DEBOUNCE = 300;  // debounce in ms
+
+const modelKitConsumer = new ModelKitConsumer();
+
+const isEmpty = (obj) => !Object.keys(obj).length;
+
+const reducer = (state, action) => {
+    // current state
+    const { searchParams, page, selectedIds, preSelected, searchResults, hasNext } = state;
+
+    // create a new state object, defaulting to the existing state
+    const newState = Object.assign({}, state);
+
+    const preSelectedIds = preSelected.map(kit => kit.id);
+
+    switch(action.type) {
+        case 'searchParam':
+            // update the search param (brand, scale or name)
+            // page is handled separately
+            const { param, value } = action;
+            if (!value) {
+                delete searchParams[param];
+            } else {
+                searchParams[param] = value;
+            }
+
+            // force creation of a new object
+            newState.searchParams = Object.assign({}, searchParams);
+            return newState;
```

Comment:
This approach to working with nested objects in reducers is prone to errors and stale state. As a general rule avoid mutating objects, even if you're copying them. Plus all the object assignment methods return a shallow copy and it's easy to run into bugs this way. 
Generally it's better to create new state in each `case` and return it:
```javascript
  case "searchParam":
        const { param, value } = action;
            if (!value) {
                const { [param]: value, ...rest } = state.searchParams;
                return rest;
            } else {
                return { ...state.searchParams, [param]: value };
            }
```

Reply:
hrm yeah, I see. I was trying to work around the insane verbosity, but the spreading operator helps with that.